rpc/handlers.go

```diff
@@ -1120,3 +1123,49 @@ func (h *Handler) EstimateFee(broadcastedTxns []BroadcastedTransaction, id Block
 
 	return estimates, nil
 }
+
+func (h *Handler) TraceTransaction(hash felt.Felt) (TransactionTrace, *jsonrpc.Error) {
+	_, _, blockNumber, err := h.bcReader.Receipt(&hash)
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+	tx, err := h.bcReader.TransactionByHash(&hash)
+	if err != nil {
+		return nil, jsonrpc.Err(jsonrpc.InternalError, err.Error())
+	}
+
+	id := &BlockID{Number: blockNumber - 1}
+
+	state, closer, err := h.stateByBlockID(id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+	defer func() {
+		if closeErr := closer(); closeErr != nil {
+			h.log.Warnw("Failed to close state in starknet_call", "err", closeErr)
+		}
+	}()
+
+	header, err := h.blockHeaderByID(id)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	var classes []core.Class
+	if tx, ok := tx.(*core.DeclareTransaction); ok {
+		class, stateErr := state.Class(tx.ClassHash)
+		if stateErr != nil {
+			return nil, jsonrpc.Err(jsonrpc.InternalError, stateErr.Error())
+		}
+		classes = append(classes, class.Class)
+	}
+
+	_, info, err := h.vm.Execute([]core.Transaction{tx}, classes, header.Number, header.Timestamp, header.SequencerAddress, state, h.network)
```

Comment:
You have to execute the entire block, not just the `txn`. So you need fetch the entire block with `blockByID`

Reply:
Fixed, but now it doesn't work :) the problem that if block contains transaction type that unsupported by `vm.marshalTxn(...)`  then entire execution is failed. Could you please explain why do I need entire block to execute? because a transaction that I requested can rely on another in the same block?

Reply:
> because a transaction that I requested can rely on another in the same block?

Yes, exactly.

> if block contains transaction type that unsupported by vm.marshalTxn(...)

Let's add support for them then:)

Reply:
It was easier than I expected 😄